src/econometrics.py

```diff
@@ -0,0 +1,240 @@
+    df['dens_change'] = df['popdensity2019'] - df['popdensity2014']
+    return df
+
+def group_and_calculate_diff(df, group_cols):
+    df = df.sort_values(by=group_cols + ['Date'])
+    grouped = df.groupby(group_cols)
+    df['diff_prixM2'] = grouped['prixM2'].diff().fillna(np.nan)
+    df['diff2_prixM2'] = grouped['prixM2'].diff().diff()
+    df['prixM2_FE'] = grouped['prixM2'].transform(lambda x: x - x.mean()) #Within transformation
+    df['year'] = df['Date'].str[:4]
+    df['diff2_prop_maison'] = grouped['prop_maison'].diff().fillna(np.nan)
+    df['lag_ir'] = grouped['ir'].shift(1)
+    df['lag2_ir'] = grouped['ir'].shift(2)
+    df['lag4_ir'] = grouped['ir'].shift(4)
+    df['diff_lag_ir'] = grouped['lag_ir'].diff().fillna(np.nan)
+    df['diff2_lag_ir'] = grouped['lag_ir'].diff().diff().fillna(np.nan)
+    return df.reset_index(drop=True)
+
+def run_regression(df, formula):
+    model = smf.ols(formula, data=df)
+    results = model.fit()
+    return(results)
+
+def aggregate_data(df, group_cols, agg_dict):
+    agg_df = df.groupby(group_cols).agg(agg_dict).reset_index()
+    agg_df['popdensity2019'] = agg_df['P20_POP'] / agg_df['SUPERF']
+    agg_df['popdensity2014'] = agg_df['P14_POP'] / agg_df['SUPERF']
+    return agg_df.drop(columns=['P20_POP', 'P14_POP', 'SUPERF'])
+
+def corplot(df, nom):
+  plt.figure(figsize=(10, 8))
+  sns.heatmap(df, annot=True, cmap='coolwarm')
+  plt.title('Correlations intra-ville pour '+nom)
+  plt.show()
+  
+def cor_ville(df, nom):
+  df = df[df['LIB_COM']==nom]
+  df = df[["prixM2", "ir","lag_ir","lag2_ir","lag4_ir", "prop_maison"]].dropna().corr()
+  return(df)
+
+
+# Reading Data
+base_path = "C:/Users/patry/Documents/GitHub/interestnpy/data/interim/"
+immo_epci = pd.read_feather(base_path + "immo_panel_epci_py.feather")
+immo_ze = pd.read_feather(base_path + "immo_panel_ze_py.feather")
+immo_com = pd.read_feather(base_path + "immo_panel_com_py.feather")
+immo_full = pd.read_feather(base_path + "immo_panel_full_py.feather")
+unemployment = pd.read_feather(base_path + "TV_controls.feather")
+control = pd.read_feather(base_path + "TI_controls.feather")
+irflation = pd.read_feather(base_path + "irflation.feather")
+
+# Aggregate data
+agg_dict_control = {
+    'Q219': 'mean', 'Q214': 'mean',
+    'P20_POP': 'sum', 'P14_POP': 'sum',
+    'SUPERF': 'sum', 'Physicist_access': 'mean',
+    'assault_for_1000': 'mean'
+}
+control_ze = aggregate_data(control, ['ZE', 'LIB_ZE'], agg_dict_control)
+control_epci = aggregate_data(control, ['EPCI', 'LIB_EPCI'], agg_dict_control)
+
+# Processing for immo_reg_com
+immo_reg_com = merge_and_transform(immo_com, irflation, control, 'COM')
+immo_reg_com_diff = group_and_calculate_diff(immo_reg_com, ['COM'])
+
+# Total 
+immo_reg_corr =immo_reg_com_diff[["prixM2", "ir","lag_ir","lag2_ir","lag4_ir","popdensity2019", "med_change", "Physicist_access", "Q219", "prop_maison", "assault_for_1000", "dens_change"]].dropna().corr()
+corplot(immo_reg_corr)
```

Comment:
There is an error message here : corplot() is missing one argument : 'nom'

Reply:
indeed! good catch